packages/flutter/lib/src/painting/image_stream.dart

```diff
@@ -196,54 +227,124 @@ abstract class ImageStreamCompleter extends Diagnosticable {
+          stack: stack,
+        );
       }
     }
+    if (_currentError != null && onError != null) {
+      try {
+          onError(_currentError.exception, _currentError.stack);
+        } catch (exception, stack) {
+          FlutterError.reportError(
+            new FlutterErrorDetails(
+              exception: exception,
+              library: 'image resource service',
+              context: 'by a synchronously-called image error listener',
+              stack: stack,
+            ),
+          );
+        }
+    }
   }
 
-  /// Stop listening for new concrete [ImageInfo] objects.
+  /// Stop listening for new concrete [ImageInfo] objects and errors from
+  /// its associated [ImageErrorListener].
   void removeListener(ImageListener listener) {
-    _listeners.remove(listener);
+    _listeners.removeWhere((_ImageListenerPair listenerPair) {
+      return listenerPair.listener == listener;
+    });
   }
 
   /// Calls all the registered listeners to notify them of a new image.
   @protected
   void setImage(ImageInfo image) {
-    _current = image;
+    _currentImage = image;
     if (_listeners.isEmpty)
       return;
-    final List<ImageListener> localListeners = new List<ImageListener>.from(_listeners);
+    final List<ImageListener> localListeners = _listeners.map<ImageListener>(
+      (_ImageListenerPair listenerPair) => listenerPair.listener
+    ).toList();
     for (ImageListener listener in localListeners) {
       try {
         listener(image, false);
       } catch (exception, stack) {
-        _handleImageError('by an image listener', exception, stack);
+        reportError(
+          context: 'by an image listener',
+          exception: exception,
+          stack: stack,
+        );
       }
     }
   }
 
-  void _handleImageError(String context, dynamic exception, dynamic stack) {
-    FlutterError.reportError(new FlutterErrorDetails(
+  /// Calls all the registered error listeners to notify them of an error that
+  /// occurred while resolving the image.
+  ///
+  /// If the same error listener is attached with multiple listeners, that
+  /// error listener will only be notified once.
+  ///
+  /// If no error listeners are attached, a [FlutterError] will be reported
+  /// instead.
+  @protected
+  void reportError({
+    String context,
+    dynamic exception,
+    StackTrace stack,
+    InformationCollector informationCollector,
+    bool silent = false,
+  }) {
+    _currentError = new FlutterErrorDetails(
       exception: exception,
       stack: stack,
       library: 'image resource service',
-      context: context
-    ));
+      context: context,
+      informationCollector: informationCollector,
+      silent: silent,
+    );
+
+    // Many listeners can have the same error listener. De-duplicate.
+    final Set<ImageErrorListener> localErrorListeners =
+        _listeners.map<ImageErrorListener>(
+          (_ImageListenerPair listenerPair) => listenerPair.errorListener
+        ).toSet()
```

Comment:
Following offline discussion, we probably want to treat the listeners and error listeners the same (not de-dup)